### DISCUSSION TITLE
Spell checking for more than one language

### DISCUSSION THREAD

**Discussion by mery6299** [zim-desktop-wiki/zim-desktop-wiki#2188] [Q&A] [answered]
How can 2 languages be handled at once?
I noticed that only one language can be spell checked at once with a plugin. What I would like to do is to spell check language 1 and ignore errors from words being correct in language 2. Is that possible?

**Comment by dehesselle** [comment, marked as answer]
I don't know any program that allows more than one language being active for spell checking.

**Reply by introt** [reply]
> Is that possible?

Not currently, no. In theory, yes.

**Comment by Crissium** [comment]
> I don't know any program that allows more than one language being active for spell checking.

Well I know one: [a spellchecking extension for VS Code](https://marketplace.visualstudio.com/items?itemName=ban.spellright). Here's a screenshot of it checking both English and French:

![image](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/91039086/a9771ab7-77ae-40bd-924c-222b03c1cb15)

It would be extremely helpful for users who regularly write in more than one language. If the functionality of spellchecking multilingual documents fits into zim, I'd gladly help with the implementation.

Thanks for considering this!

**Reply by dehesselle** [reply]
Interesting find! But I still think that this kind of functionality is a corner case - not that it's not useful but that I don't expect it to be easily implemented/integrated into the existing systems. No way to know for sure without doing the necessary digging into the source code, which likely will include GTK and its interface to spell checking libraries. If that's the kind of help you can offer, start looking into this.

Generally speaking, you can expect acceptance of such a feature will mostly be based on maintainability, i.e. "does it require hundreds of lines of code of ugly hacks and workarounds" to make it work (see [measurement of code quality](https://www.reddit.com/r/ProgrammerHumor/comments/7ym547/the_only_valid_measurement_of_code_quality_is/)). As I said above: you won't know until you start digging. Or maybe ask the guy who wrote [`gedit`](https://apps.gnome.org/en/TextEditor/) as a first step?

**Reply by Crissium** [reply]
After some experimenting I think directly implementing this in zim is too messy and simply won't work, as you have suggested. So I think perhaps I should try first modifying one of the spellchecking libraries zim use, since the spellchecking plugin is just a thin wrapper.

**Comment by Crissium** [comment]
So eventually I created my own plugin to implement this. For anyone interested, it is published [here](https://github.com/Crissium/zim-multilingual-spellchecker).

![Screenshot](https://github.com/zim-desktop-wiki/zim-desktop-wiki/assets/91039086/91196423-829a-4a20-b508-0b8df8e64d1a)

**Reply by introt** [reply]
Nice! Please add this under https://github.com/jaap-karssenberg/zim-wiki/wiki/Plugins

**Reply by Simplenuity** [reply]
Thanks, @Crissium! Very helpful, indeed!

**Reply by Crissium** [reply]
Thanks for the kind words!

I have already included it in the wiki, but I must admit that the plug-in was developed in a hurry and a lot of code was just blindly copied over from the original one. I didn't even bother to get rid of the Adapter pattern (for the three spellchecking libraries). I'll try to polish up the code next week.
